### ISSUE TITLE
hledger roi results in -100% or 0% TWR for finished investments with positive rate of return

### ISSUE THREAD

**Issue by LaurenzWiskott2** [simonmichael/hledger#1791]
Given the file `hledger.ledger`
```
2010-01-01   A1
    savingA                      100.00 €
    checking   

2011-12-31   A2
    savingA                        0.00 €
    dividendA                     -8.00 €
    checking
    
2012-03-29   A3
    savingA                     -100.00 €
    checking
    
# ========================================

2010-01-01   B1
    savingB                      100.00 €
    checking   

2011-12-31   B2
    savingB                        0.00 €
    dividendB                     -8.00 €
    checking
    
2012-03-29   B3
    saving                      -100.00 €
    checking
```
the command `hledger -f hledger.ledger roi --inv saving --pnl dividend` leads to
```
+---++------------+------------++---------------+----------+-------------+---------++-------+----------+
|   ||      Begin |        End || Value (begin) | Cashflow | Value (end) |     PnL ||   IRR |      TWR |
+===++============+============++===============+==========+=============+=========++=======+==========+
| 1 || 2010-01-01 | 2012-03-29 ||             0 | -16.00 € |           0 | 16.00 € || 3.52% | -100.00% |
+---++------------+------------++---------------+----------+-------------+---------++-------+----------+
```
and `hledger -f hledger.ledger roi --inv savingA --pnl dividendA` leads to
```
+---++------------+------------++---------------+----------+-------------+--------++-------+-------+
|   ||      Begin |        End || Value (begin) | Cashflow | Value (end) |    PnL ||   IRR |   TWR |
+===++============+============++===============+==========+=============+========++=======+=======+
| 1 || 2010-01-01 | 2012-03-29 ||             0 |  -8.00 € |           0 | 8.00 € || 3.52% | 0.00% |
+---++------------+------------++---------------+----------+-------------+--------++-------+-------+
```

This looks like a bug to me. 
Firstly, why is the TWR different in the two cases, since the conditions are pretty much identical, just doubled?
Secondly, why is the TWR not positive and even negative in the first case, even though there is definitely a positive rate of return by any measure I think?

This might be related to another slightly irritating behavior. Given the file `hledger2.ledger`
```
2010-12-01   A1
    savingA                      100.00 €
    checking   

2010-12-31   A2
    savingA                        0.00 €
    dividendA                     -8.00 €
    checking
    
2011-03-29   A3
    savingA                     -100.00 €
    checking
```
the command `hledger -f hledger2.ledger roi --inv savingA --pnl dividendA -Y` leads to
```
+---++------------+------------++---------------+-----------+-------------+--------++---------+-------+
|   ||      Begin |        End || Value (begin) |  Cashflow | Value (end) |    PnL ||     IRR |   TWR |
+===++============+============++===============+===========+=============+========++=========+=======+
| 1 || 2010-01-01 | 2010-12-31 ||             0 |   92.00 € |    100.00 € | 8.00 € || 148.02% | 8.00% |
| 2 || 2011-01-01 | 2011-12-31 ||      100.00 € | -100.00 € |           0 |      0 ||   0.00% | 0.00% |
+---++------------+------------++---------------+-----------+-------------+--------++---------+-------+
```
In this case there is no money at all in the savingA account until December, so there is no evidence whatsoever how much rate of return there is in this time. `hledger roi' seems to treat times without any money in the account, i.e. times without any evidence, as if there were no rate of return, i.e. 0% TWR. I find this inappropriate. I think such times should just be ignored rather than assuming TWR = 0%.

Merry Christmas and thanks for this great software.

**Comment by LaurenzWiskott2** [comment]
PS: I used `hledger 47b5b6b, linux-x86_64`, but I get the same behavior with `hledger 1.24.1-g7799d526b-20211210` downloaded just a few minutes ago.

**Comment by adept** [comment]
@LaurenzWiskott2 thank you, -100% is definitely a bug, and will be fixed by #1794 . Likewise, 0% reported by your second example is also a bug and will be fixed by the same PR (and TWR reported would be 3.49% , as it should've been)

As for the last point (regarding the period of time with no info), this is specifically how TWR is supposed to behave. Unlike IRR/XIRR, the length of the time interval between PnL or cashflow events is completely ignored in TWR computations, which presents an issue for the periodic reporting. As manual says, all TWR rates would be annualized during reporting, so it looks to me that 8% is correct in this case
